File: 0279-Perfect_Squares/Perfect_Squares.java
class Solution {
    public int numSquares(int n) {
        int[] arr = new int[n + 1];
        
        for(int i = 1; i <= n; i++){
            int least = Integer.MAX_VALUE;
            for(int j = 1; (j * j) <= i; j++){
                least = Math.min(least, arr[i - (j*j)] + 1);
            }
            arr[i] = least;
        }
        
        return arr[n];
    }
}

File: 0279-PerfectSquares/PerfectSquares.go
package perfectsquares

func numSquares(n int) int {
	dp := make([]int, n+1)

	for i := 1; i <= n; i++ {
		dp[i] = 100001
		for j := 1; j*j <= i; j++ {
			dp[i] = min(dp[i], dp[i-j*j]+1)
		}
	}

	return dp[n]
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}
